Add mistral chat adapter, update Gemma 2 in klite

Nexesenex opened the pull request.

Nexesenex committed:
Add mistral chat adapter

pi6am committed:
Add the DRY dynamic N-gram anti-repetition sampler

The DRY (Do not Repeat Yourself) sampler is a dynamic N-gram
repetition penalty that negatively scores tokens that would extend
sequences that already appear in the context.

See this discussion for a motivation and explanation of the sampler:
https://github.com/oobabooga/text-generation-webui/pull/5677

This implementation of DRY mostly aligns with the obabooga version
with a few modifications. It uses a more efficient linear scanning
algorithm to identify repetitions. It also supports multi-token
sequence breakers. As a limitation, this implementation reuses
the rep pen range parameter, rather than introducing a new range
just for the DRY sampler.

There is a separate change to lite.koboldai.net that exposes the DRY
sampler parameters to KoboldAI Lite, so none of the embed files have
been changed as part of this commit.

LostRuins commented:
Can you PR the changes for lite separately to https://github.com/LostRuins/lite.koboldai.net/ dev branch? The klite here will get overwritten, nothing modified to it will stay

Nexesenex commented:
All done.

pi6am committed:
Update default DRY parameters to match lite

pi6am committed:
Improve DRY token debug logging

pi6am committed:
Merge pull request #1 from LostRuins/concedo

Merge upstream

pi6am committed:
Merge branch 'concedo' into feat/dry-sampler

pi6am committed:
Replace `and` with `&&` to fix MSVC compile error

Little known fact: The C++98 standard defines `and` as an
alternative token for the `&&` operator (along with a bunch
of other digraphs). MSVC does not allow these without using
the /Za option or including the <iso646.h> header. Change to
the more standard operator to make this code more portable.

pi6am committed:
Fix MSVC compile error because log is not constexpr

Replace the compile-time computation with a floating-point
approximation of log(std::numeric_limits<float>::max()).

Nexesenex committed:
Merge pull request #225 from pi6am/feat/dry-sampler

Feat/dry sampler

Nexesenex commented:
Oh lol. I bungled it. I'm closing this. xD
